### PR TITLE
[XLA:GPU] Handle fp16 batchnorm to thunk to cudnn batchnorm half kernels

### PR DESCRIPTION
**Pull request by AyanmoI**
When using cuDNN for batchnormalization, xla currently thunks fp16 batchnorms to fp32 kernels by casting fp16 inputs and gradients to fp32 in the tf2xla bridge. This is non-optimal since cuDNN 7.6 has highly efficient `Half` kernels that can be leveraged. This PR modifies `cudnn_batchnorm_rewriter` such that the converts introduced in the bridge get eliminated and the corresponding fp16 cudnn APIs get called (changes for the thunk already merged https://github.com/tensorflow/tensorflow/pull/32887)
Note that instead of eliminating convert ops, I add converts which then get eliminated by `algebraic_simplifier`. PR https://github.com/tensorflow/tensorflow/pull/32435  adding capability to eliminate of convert pairs was merged earlier.

### CODE REVIEW

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc [hunk 1]
```diff
@@ -78,13 +91,28 @@ Status Visitor::HandleBatchNormInference(HloInstruction* batch_norm) {
 
   std::vector<HloInstruction*> operands(batch_norm->operands().begin(),
                                         batch_norm->operands().end());
+
+  if (IsF32BatchNormWithFP16Inputs(batch_norm)) {
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Maybe just remember this in a local variable?

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc [hunk 2]
```diff
@@ -143,17 +185,33 @@ Status Visitor::HandleBatchNormTraining(HloInstruction* batch_norm) {
...
+    new_gte = AddConvert(new_gte, F32);
+  }
+  // Repackage the results. Athough this tuple is redundant when convert is not
+  // inserted, TupleSimplifier eliminates the Tuple eventually
+  std::unique_ptr<HloInstruction> replacing_tuple = HloInstruction::CreateTuple(
+      {new_gte,
+       computation_->AddInstruction(HloInstruction::CreateGetTupleElement(
+           new_tuple->shape().tuple_shapes(1), new_tuple, 1)),
+       computation_->AddInstruction(HloInstruction::CreateGetTupleElement(
+           new_tuple->shape().tuple_shapes(1), new_tuple, 2))});
```
Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Should this be `new_tuple->shape().tuple_shapes(2)`?

Comment (sanjoy) [inline tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Is all of this repackaging necessary or can we directly create the final tuple using GTE's on `libcall`, `variance` etc.?

Comment (sanjoy) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Same comment for `HandleBatchNormGrad`.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc (hunk 1 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: I actually had it as a local variable in the earlier PR but changed it based on your comment :). 
Do u want something like this? (See line 99 onwards in the modified version) https://github.com/tensorflow/tensorflow/pull/32665/commits/5e8de43c588cee0ac510e35cfb62e91c734cca61#diff-f57e826cf2d3f0a9699cb0451fa78738

Comment (sanjoy) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Yes, but we need a more targeted name than `is_convert_inserted` since the fundamental property is that this is an fp16 batchnorm, not that a convert was inserted.

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc (hunk 2 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: You mean `new_tuple` could have been bypassed..right? Agreed! I initially had it here to signify that the instructions between `new_tuple` and `replacing_tuple` correspond to handling F16 batchnorm. I agree that it doesn't improve anything in terms of readability or making the code more understandable. Note that `HandleBatchNormGrad` doesn't have any additional tuples. The comments `// Repackage the results. Athough ...` implies that instead of adding more code here to not have another GTE for F32, I chose to have one and let the optimizer eliminate the redundant tuple.

Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: I will modify `HandleBatchNormTraining` but I don't think `HandleBatchNormGrad` needs changes unless I am misinterpreting your comment :).

Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Correct..Sorry that is a bug

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc (hunk 1 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: I see...I read your earlier comment incorrectly. Sorry abt that.

**Comment by AyanmoI** [comment]
Hi @sanjoy. I made the changes. Please let me know if they are ok.

File: tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc [hunk 3]
```diff
@@ -78,13 +91,30 @@ Status Visitor::HandleBatchNormInference(HloInstruction* batch_norm) {
 
   std::vector<HloInstruction*> operands(batch_norm->operands().begin(),
                                         batch_norm->operands().end());
+
+  bool is_batchnorm_with_fp16_inputs = false;
+  if (IsF32BatchNormWithFP16Inputs(batch_norm)) {
```
[the remark this answers is not in the document]
Comment (sanjoy) [reply tensorflow/compiler/xla/service/gpu/cudnn_batchnorm_rewriter.cc]: Same below.

File: tensorflow/compiler/xla/debug_options_flags.cc [hunk 4]
```diff
@@ -42,9 +42,8 @@ DebugOptions DefaultDebugOptionsIgnoringFlags() {
   opts.set_xla_cpu_use_mkl_dnn(true);
 #endif  // INTEL_MKL
   opts.set_xla_gpu_max_kernel_unroll_factor(4);
-  // Set cudnn batchnorm off by default; it does not provide a performance win
-  // on average.
-  opts.set_xla_gpu_use_cudnn_batchnorm(false);
+  // Set cudnn batchnorm on by default.
+  opts.set_xla_gpu_use_cudnn_batchnorm(true);
```
Comment (sanjoy) [inline tensorflow/compiler/xla/debug_options_flags.cc]: Hi, do you mind splitting out this part of the change?  That way we can land this bit without any performance implications and the test the performance on our internal benchmarks without risking a regression.

Comment (sanjoy) [reply tensorflow/compiler/xla/debug_options_flags.cc]: You can probably set this explicitly in the unit tests so that we know we're testing the right thing.

**Review by sanjoy — CHANGES_REQUESTED** [review]

File: tensorflow/compiler/xla/debug_options_flags.cc (hunk 4 above)
Comment (AyanmoI) [reply tensorflow/compiler/xla/debug_options_flags.cc]: Sure I can split this. Actually the tests always explicitly runs both the cudnn and batchnorm-expanded+fusion version irrespective of what the default is set here. I had the same concern but then I verified this.  The test sets `use_cudnn_batchnorm` parameter during construction to `false` and then flips it . See https://github.com/tensorflow/tensorflow/blob/07b5ca443887fd6d739ef6055e6cd3f2c02ba226/tensorflow/compiler/xla/tests/batch_normalization_test.cc#L95 (`class BatchNormalizationTest`). Similar thing happens for the other `class BatchNormTestManySizes` (https://github.com/tensorflow/tensorflow/blob/07b5ca443887fd6d739ef6055e6cd3f2c02ba226/tensorflow/compiler/xla/tests/batch_normalization_test.cc#L405)

**Comment by AyanmoI** [comment]
@sanjoy I have addressed your comments and moved the flag setting change to https://github.com/tensorflow/tensorflow/pull/33583. Please take a look.

**Review by sanjoy — APPROVED** [review]
looks great, thanks!

**Comment by AyanmoI** [comment]
Hi @sanjoy ...this PR seems to be stuck at this stage...any issues with merging this?

**Comment by cheshire** [comment]
> xla currently thunks fp16 batchnorms to fp32 kernels by casting fp16 inputs and gradients to fp32 in the tf2xla bridge

Naive question: should we do it in the bridge in the first place?

**Comment by cheshire** [comment]
> Naive question: should we do it in the bridge in the first place?

Actually nevermind, I misunderstood the lowering.
